src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -970,6 +979,407 @@ public void recursivelyDelete(final Project project, final boolean commitIndex)
         }
     }
 
+    /**
+     * Deletes a list of Projects (identified by their UUIDs) and all objects dependent on them.
+     * @param uuids the UUIDs of the Projects to delete
+     *
+     * NB: if ON DELETE rules had been set up, this would be as simple as "delete from project where ..."
+     */
+    @Override
+    public void deleteProjectsByUUIDs(Collection<UUID> uuids) {
+        final var errorByUUID = new HashMap<String, String>();
+
+        for (Iterator<UUID> it = uuids.iterator(); it.hasNext(); ) {
+            UUID uuid = it.next();
+            final Project project = getObjectByUuid(Project.class, uuid);
+            if (project == null) {
+                errorByUUID.put(String.valueOf(uuid), "Project not found");
+                it.remove();
+            }
+            else {
+                if (!hasAccess(principal, project)) {
+                    errorByUUID.put(String.valueOf(uuid), "Access denied to project");
+                    it.remove();
+                }
+            }
+        }
+        if (!errorByUUID.isEmpty()) {
+            throw ProjectOperationException.forDeletion(errorByUUID);
+        }
+
+        String[] uuidsArray = uuids.stream().map(UUID::toString).toArray(String[]::new);
+        String commaSeparatedUUIDs = String.join(",", Collections.nCopies(uuidsArray.length, "?"));
+        var queryParameter = Config.getInstance().getProperty(Config.AlpineKey.DATABASE_DRIVER).equals("com.microsoft.sqlserver.jdbc.SQLServerDriver") ?
+                commaSeparatedUUIDs : uuidsArray;
+        String inExpression = Config.getInstance().getProperty(Config.AlpineKey.DATABASE_DRIVER).equals("com.microsoft.sqlserver.jdbc.SQLServerDriver") ?
+            "IN(SELECT value FROM STRING_SPLIT(?, ','))" : "= ANY(?)";
+
+        runInTransaction(() -> {
+            try (var ignored = new ScopedCustomization(pm).withProperty(PROPERTY_QUERY_SQL_ALLOWALL, "true")) {
+                Query<?> sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "PROJECTMETRICS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
```

Comment:
Doing one initial `select id from project where uuid = any(:uuids)` query to retrieve the projects' IDs would allow you to omit a lot of subqueries like this.

Reply:
Now implemented basically as suggested, and it seems the performance increased further by doing so: those ~11000 projects now get deleted in ~17 seconds instead of 25 seconds. Probably mostly thanks to loading all projects with one DB round-trip instead of 1 hit per project.